Refactor virtual thread inspector access

fengxue-IS opened the pull request:
Add carrier J9VMThread info to isSuspendedInternal field
Fix getVMThread API to account for suspended virtual thread under transition

Fix: #18504

fengxue-IS commented:
[personal build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/21074/) in progress
@babsingh can you take a quick look at this?
local testing on #18735 test passed 30/30, I will launch a grinder to verify once the personal build passes

babsingh commented on runtime/jvmti/suspendhelper.cpp:
```diff
@@ -49,7 +49,7 @@ suspendThread(J9VMThread *currentThread, jthread thread, BOOLEAN allowNull, BOOL
 	 * can be taken for both platform and virtual threads.
 	 */
-	if (0 != J9OBJECT_U32_LOAD(currentThread, threadObject, vm->isSuspendedInternalOffset)) {
+	if (J9_ARE_ANY_BITS_SET(J9OBJECT_U64_LOAD(currentThread, threadObject, vm->isSuspendedInternalOffset), 1)) {
```
for better readability, nested calls should be avoided. this sequence is repeated multiple times, a helper should be added for it.
````
U_64 suspendedInternal = J9OBJECT_U64_LOAD(currentThread, threadObject, vm->isSuspendedInternalOffset);
if (J9_ARE_ANY_BITS_SET(suspendedInternal, 0x1)) {

babsingh commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -570,17 +586,34 @@ releaseVThreadInspector(J9VMThread *currentThread, jobject thread)
-	) {
-		/* Field updated by another thread, try again. */
-		vthreadInspectorCount = J9OBJECT_I64_LOAD(currentThread, threadObj, vm->virtualThreadInspectorCountOffset);
+	U_64 isSuspendedFlag = J9OBJECT_U64_LOAD(currentThread, threadObj, vm->isSuspendedInternalOffset);
```
`isSuspendedFlag` or `isSuspendedInternalOffset`: it is no longer a boolean; so, these names should be updated to reflect the new behaviour.

babsingh commented on runtime/jvmti/jvmtiThread.cpp:
```diff
@@ -1382,27 +1397,32 @@ jvmtiSuspendResumeCallBack(J9VMThread *vmThread, J9MM_IterateObjectDescriptor *o
 			J9VMThread *targetThread = NULL;
 			j9object_t carrierThread = (j9object_t)J9VMJAVALANGVIRTUALTHREAD_CARRIERTHREAD(vmThread, vthread);
+			U_64 isSuspendedFlag = J9OBJECT_U64_LOAD(vmThread, vthread, vm->isSuspendedInternalOffset);
+			BOOLEAN inTransition = J9_ARE_ANY_BITS_SET(isSuspendedFlag, ~0x1);
```
using 0x1 makes it difficult to understand the logic. we should introduce macros and helper methods similar to what we do for `Continuation.state`.
```
#define J9_VIRTUALTHREAD_INTERNAL_STATE_SUSPENDED 0x1
#define J9_VIRTUALTHREAD_INTERNAL_STATE_CARRIERID_MASK (~(uintptr_t)0xff)
```

babsingh commented on runtime/jvmti/jvmtiStackFrame.c:
```diff
@@ -465,7 +465,7 @@ jvmtiPopFrame(jvmtiEnv* env,
 			if ((currentThread != targetThread)
 #if JAVA_SPEC_VERSION >= 21
-			&& (0 == J9OBJECT_U32_LOAD(currentThread, threadObject, vm->isSuspendedInternalOffset))
+			&& (J9_ARE_NO_BITS_SET(J9OBJECT_U64_LOAD(currentThread, threadObject, vm->isSuspendedInternalOffset), 1))
```
we should add helper methods for such checks; in the future, we will only need to update the helper method, instead of updating all the instances.

babsingh commented on runtime/jvmti/jvmtiHelpers.cpp:
```diff
@@ -168,6 +168,13 @@ getVMThread(J9VMThread *currentThread, jthread thread, J9VMThread **vmThreadPtr,
 		if (NULL != carrierThread) {
 			targetThread = J9VMJAVALANGTHREAD_THREADREF(currentThread, carrierThread);
 		}
+		if (-1 > J9OBJECT_I64_LOAD(currentThread, threadObject, vm->virtualThreadInspectorCountOffset)) {
+			/* If the virtual thread is suspended in transition, check if the mounting process is already completed. */
+			J9VMThread *carrierVMThread = (J9VMThread *)(J9OBJECT_U64_LOAD(currentThread, threadObject, vm->isSuspendedInternalOffset) & ~0x1);
+			if (NULL != carrierVMThread->currentContinuation) {
+				targetThread = carrierVMThread;
+			}
+		}
```
This shouldn't be possible since an inTransition suspended thread should have the halt flag set.

fengxue-IS replied:
if the vthread is suspended after entering critical section, then it is possible that the carrierThread field on vthread have not been set, but the halt flag is set on the carrier as it is halted by the `internalSuspendState`. so the only access to carrier vm thread would be through the `internalSuspendState` field

babsingh commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -274,21 +274,57 @@ enterVThreadTransitionCritical(J9VMThread *currentThread, jobject thread)
+		vthread = J9_JNI_UNWRAP_REFERENCE(thread);
+	}
+
+	/* Update to virtualThreadInspectorCount must be after clearing isSuspendedInternal field to avoid allowing */
```
Comment is incomplete.

babsingh commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -274,21 +274,57 @@ enterVThreadTransitionCritical(J9VMThread *currentThread, jobject thread)
+	while (!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(currentThread, vthread, vm->isSuspendedInternalOffset, (U_64)currentThread, 0)) {
+		/* Wait if the suspend flag is set. */
+		vmFuncs->internalReleaseVMAccess(currentThread);
+		VM_AtomicSupport::yieldCPU();
```
similar to the `enterVThreadTransitionCritical`, we should use here `f_threadSleep` since it is more granular and it will reduce CPU hogging.

babsingh commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -274,21 +274,57 @@ enterVThreadTransitionCritical(J9VMThread *currentThread, jobject thread)
+	/* Link the current J9VMThread with the virtual thread object. */
+	if (!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(currentThread, threadObj, vm->isSuspendedInternalOffset, 0, (U_64)currentThread)) {
+		/* If virtual thread is suspended while unmounted, reset the inspectorCount and do a wait and retry. */
+		if (J9_ARE_ANY_BITS_SET(J9OBJECT_U64_LOAD(currentThread, threadObj, vm->isSuspendedInternalOffset), 1)) {
+			J9OBJECT_I64_STORE(currentThread, threadObj, vm->virtualThreadInspectorCountOffset, 0);
+		}
```
Why do we need to reset the `inspectorCount`?

fengxue-IS replied:
if a vthread is suspended while unmounted, there will be no halt flag since the native vm thread is not linked.
when that thread is attempted to be mounted, we need to avoid acquiring either the inspector count or linking to vmthread so it is still viewed as a unmounted thread

babsingh reviewed (commented):
first review pass.

fengxue-IS commented on runtime/jvmti/jvmtiThread.cpp:
```diff
@@ -1179,11 +1179,27 @@ resumeThread(J9VMThread *currentThread, jthread thread)
+			if (inTransition) {
+				targetThread = (J9VMThread*)(isSuspendedFlag & ~0x1);
+				/* For virtual thread in transition, ensure thread cannot start running until releaseVMThread is called. */
+				vm->internalVMFunctions->haltThreadForInspection(currentThread, targetThread);
```
since the target thread is already halted by suspend flag, `haltThreadForInspection` should directly return after setting the inspection flag, though i guess it wouldn't hurt to do the refetch.

fengxue-IS commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -570,17 +586,34 @@ releaseVThreadInspector(J9VMThread *currentThread, jobject thread)
+	U_64 isSuspendedFlag = J9OBJECT_U64_LOAD(currentThread, threadObj, vm->isSuspendedInternalOffset);
+
+	/* If a thread is suspended, then access maybe allowed during transition phase with negative inspector count. */
+	if (J9_ARE_ANY_BITS_SET(isSuspendedFlag, ~0x1)) {
+		Assert_VM_true(vthreadInspectorCount < -1);
+		while (!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(
```
this checks if thread is in transition, the suspend bit is not actually checked since the caller could be resuming this thread hence removing the suspend bit before releasing inspector count. I'll update the comment to reflect this more clearly.

fengxue-IS commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -609,7 +622,7 @@ JVM_VirtualThreadHideFrames(
 	if (!hide) {
 		Assert_SC_true(hiddenFrames);
-		exitVThreadTransitionCritical(currentThread, vThreadObj);
+		exitVThreadTransitionCritical(currentThread, (jobject)&currentThread->threadObject);
```
actually this should be fine, since the jobject passed in is the address of `currentThread->threadObject` even if the `threadObject` is moved by GC, the new address should be written to the same address by GC

fengxue-IS commented:
answered questions related to code change, updated code to use constant flags and common API as suggested.

@babsingh can you take another look please

babsingh commented on runtime/jvmti/jvmtiThread.cpp:
```diff
@@ -1179,11 +1183,27 @@ resumeThread(J9VMThread *currentThread, jthread thread)
 #if JAVA_SPEC_VERSION >= 19
 		J9JavaVM *vm = currentThread->javaVM;
 		j9object_t threadObject = J9_JNI_UNWRAP_REFERENCE(thread);
+		U_64 internalSuspendState = J9OBJECT_U64_LOAD(currentThread, threadObject, vm->internalSuspendStateOffset);
+		U_64 carrierVMThread = (internalSuspendState & J9_VIRTUALTHREAD_INTERNAL_STATE_CARRIERID_MASK);
+
+		/* Update internal flag before clearing flag to avoid sync issue. */
+		if (J9_ARE_NO_BITS_SET(internalSuspendState, J9_VIRTUALTHREAD_INTERNAL_STATE_SUSPENDED)) {
+			rc = JVMTI_ERROR_THREAD_NOT_SUSPENDED;
```
the current approach only loads the field once. but, it adds another location to maintain if the code related to the bit-mask changes. if we use the helpers here, there will be an extra load, but the code will be easy to maintain since the bit-mask operations will only be done from the helpers.

```suggestion
		J9VMThread *carrierVMThread = VM_VMHelpers::getCarrierVMThread(currentThread, threadObject);

		/* Update internal flag before clearing flag to avoid sync issue. */
		if (!VM_VMHelpers::isThreadSuspended(currentThread, threadObject)) {
			rc = JVMTI_ERROR_THREAD_NOT_SUSPENDED;
```

babsingh commented on runtime/jvmti/jvmtiThread.cpp:
```diff
@@ -1382,27 +1401,32 @@ jvmtiSuspendResumeCallBack(J9VMThread *vmThread, J9MM_IterateObjectDescriptor *o
 			J9JavaVM *vm = vmThread->javaVM;
 			J9VMThread *targetThread = NULL;
 			j9object_t carrierThread = (j9object_t)J9VMJAVALANGVIRTUALTHREAD_CARRIERTHREAD(vmThread, vthread);
+			U_64 isSuspendedFlag = J9OBJECT_U64_LOAD(vmThread, vthread, vm->internalSuspendStateOffset);
+			BOOLEAN inTransition = J9_ARE_ANY_BITS_SET(isSuspendedFlag, ~0x1);
 			if (NULL != carrierThread) {
 				targetThread = J9VMJAVALANGTHREAD_THREADREF(vmThread, carrierThread);
 				Assert_JVMTI_notNull(targetThread);
+			} else if (inTransition) {
+				/* Virtual thread during transition. */
+				targetThread = (J9VMThread *)(isSuspendedFlag & ~0x1);
 			}
```
we should use the new helper here.
```suggestion
			J9VMThread *carrierVMThread = VM_VMHelpers::getCarrierVMThread(vmThread, vthread);
			BOOLEAN inTransition = (NULL != carrierVMThread);
			if (NULL != carrierThread) {
				targetThread = J9VMJAVALANGTHREAD_THREADREF(vmThread, carrierThread);
				Assert_JVMTI_notNull(targetThread);
			} else if (inTransition) {
				/* Virtual thread during transition. */
				targetThread = carrierVMThread;
			}
```

babsingh commented on runtime/jvmti/jvmtiThread.cpp:
```diff
@@ -1382,27 +1401,32 @@ jvmtiSuspendResumeCallBack(J9VMThread *vmThread, J9MM_IterateObjectDescriptor *o
 			} else {
 				/* Resume the virtual thread. */
-				if ((NULL != targetThread) && (vthread == targetThread->threadObject)) {
+				J9OBJECT_U64_STORE(vmThread, vthread, vm->internalSuspendStateOffset, (isSuspendedFlag & ~0x1));
```
we should use the new macro here.
```suggestion
				J9OBJECT_U64_STORE(vmThread, vthread, vm->internalSuspendStateOffset, (isSuspendedFlag & ~J9_VIRTUALTHREAD_INTERNAL_STATE_SUSPENDED));
```

babsingh commented on runtime/jvmti/jvmtiThread.cpp:
```diff
@@ -1382,27 +1401,32 @@ jvmtiSuspendResumeCallBack(J9VMThread *vmThread, J9MM_IterateObjectDescriptor *o
 			}
 			if (data->is_suspend) {
-				if ((NULL != targetThread) && (vthread == targetThread->threadObject)) {
+				J9OBJECT_U64_STORE(vmThread, vthread, vm->internalSuspendStateOffset, (isSuspendedFlag | 1));
```
we should use the new macro here.
```suggestion
				J9OBJECT_U64_STORE(vmThread, vthread, vm->internalSuspendStateOffset, (isSuspendedFlag | J9_VIRTUALTHREAD_INTERNAL_STATE_SUSPENDED));
```

babsingh commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -536,6 +536,22 @@ acquireVThreadInspector(J9VMThread *currentThread, jobject thread, BOOLEAN spin)
 	vthreadInspectorCount = J9OBJECT_I64_LOAD(currentThread, threadObj, vm->virtualThreadInspectorCountOffset);
 	if (vthreadInspectorCount < 0) {
+		if (VM_VMHelpers::isThreadSuspended(currentThread, threadObj)) {
+			/* Thread is suspended, so we can directly allow access to thread without waiting for transition to complete.
+			/* Decrement the inspector count to represent inspector during suspended transition.
+			 */
```
comment formatting.
```suggestion
			/* Thread is suspended, so we can directly allow access to thread without waiting for transition to complete.
			 * Decrement the inspector count to represent inspector during suspended transition.
			 */
```

babsingh reviewed (commented):
Thanks for addressing the previous comments. The changes LGTM functionally. 

re https://github.com/eclipse-openj9/openj9/issues/18735#issuecomment-1977455105: In addition to `SuspendResume2.java`, did you also run a grinder with `SelfSuspendDisablerTest.java`?

fengxue-IS commented:
> re [#18735 (comment)](https://github.com/eclipse-openj9/openj9/issues/18735#issuecomment-1977455105): In addition to `SuspendResume2.java`, did you also run a grinder with `SelfSuspendDisablerTest.java`?

No, I will launch a grinder once the new review comments have been addressed.

babsingh commented on runtime/jvmti/jvmtiForceEarlyReturn.c:
```diff
@@ -163,7 +163,7 @@ jvmtiForceEarlyReturn(jvmtiEnv* env,
 			if ((currentThread != targetThread)
 #if JAVA_SPEC_VERSION >= 21
-			&& (0 == J9OBJECT_U32_LOAD(currentThread, threadObject, vm->isSuspendedInternalOffset))
+			&& (J9_ARE_NO_BITS_SET(J9OBJECT_U64_LOAD(currentThread, threadObject, vm->internalSuspendStateOffset), J9_VIRTUALTHREAD_INTERNAL_STATE_SUSPENDED))
```
the purpose adding the helpers goes away if we are not consistently using them to ease readability and maintenance. it might be straightforward to convert these into C++ files. if not, a simple approach would be to add wrapper methods in `jvmtiHelpers.cpp` which can invoke the helper methods in `VM_VMHelpers::*`. All JVMTI code should the wrapper methods in `jvmtiHelpers.cpp`, which should be inlined by the compiler. in the long run, we can transform the `.c` files into `.cpp` as part of our C++ transformation effort.

babsingh replied:
> i also noticed a few other locations where the new helper can be used.

this applies to jvmtiLocalVariable.c and jvmtiStackFrame.c

also, jvmtiForceEarlyReturn.cpp requires a new line at the end of the file.

babsingh replied:
> I've converted the file into cpp, local build/testing on JDK 22 passed, will launch a personal build with 8/11 to verify this didn't break any other JDK level

the cpp conversion changes look safe. i will launch the builds in the PR to save time.

fengxue-IS commented:
Launched [new personal build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/21153/) to verify JDK 8/11 after converting file to cpp.

Addressed all review comments. FYI @babsingh will launch the grinder once the personal build passed

fengxue-IS commented:
Updated jvmtiLocalVariable.c and jvmtiStackFrame.c to cpp as well, corrected compiling warning/errors on JDK22, local build/testing passed on xlinux w/ gcc11.4.0.

babsingh commented:
Code LGTM. Can you squash the commits? I will start the PR builds after the commits are squashed.

fengxue-IS commented:
squashed commits, updated PR description

babsingh commented:
jenkins test sanity.functional,extended.functional win,plinux jdk8,jdk11

babsingh commented:
jenkins test sanity.functional,extended.functional,sanity.openjdk,extended.openjdk zlinux,win,aix jdk21

babsingh commented:
jenkins test extended.functional,sanity.openjdk,extended.openjdk alinux jdk22

keithc-ca commented:
jenkins test sanity.functional,extended.functional,sanity.openjdk,extended.openjdk aix jdk21

babsingh commented on runtime/oti/VMHelpers.hpp:
```diff
@@ -2109,6 +2109,20 @@ class VM_VMHelpers
 		}
 	}
+
+	static VMINLINE BOOLEAN
+	isThreadSuspended(J9VMThread *currentThread, j9object_t thread)
+	{
+		U_64 internalSuspendState = J9OBJECT_U64_LOAD(currentThread, thread, currentThread->javaVM->internalSuspendStateOffset);
+		return J9_ARE_ANY_BITS_SET(internalSuspendState, J9_VIRTUALTHREAD_INTERNAL_STATE_SUSPENDED);
+	}
+
+	static VMINLINE J9VMThread *
+	getCarrierVMThread(J9VMThread *currentThread, j9object_t thread)
+	{
+		U_64 internalSuspendState = J9OBJECT_U64_LOAD(currentThread, thread, currentThread->javaVM->internalSuspendStateOffset);
+		return (J9VMThread *)(internalSuspendState & J9_VIRTUALTHREAD_INTERNAL_STATE_CARRIERID_MASK);
+	}
```
Function description is missing. It is also missing for the two methods above. All other methods next to them have a function description.

babsingh commented:
PR builds:
- JDK21 aix: https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/5214
- JDK21 zlinux, win: https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/5208
- JDK22 alinux: https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/5209
- JDK 8,11 plinux, win: https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/5207

The Windows PR builds failed due to infra issues. I didn't see any functional failures in them.

An unrelated and known failure is seen in the AIX extended.openjdk build:
```
java.lang.Error: Probable fatal error: No physical fonts found.
	at java.desktop/sun.font.SunFontManager.lambda$getDefaultPhysicalFont$0(SunFontManager.java:1034)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at java.desktop/sun.font.SunFontManager.getDefaultPhysicalFont(SunFontManager.java:1034)
	at java.desktop/sun.font.CompositeFont.doDeferredInitialisation(CompositeFont.java:291)
	at java.desktop/sun.font.CompositeFont.getSlotFont(CompositeFont.java:376)
	at java.desktop/sun.font.CompositeGlyphMapper.initMapper(CompositeGlyphMapper.java:81)
	at java.desktop/sun.font.CompositeGlyphMapper.<init>(CompositeGlyphMapper.java:62)
	at java.desktop/sun.font.CompositeFont.getMapper(CompositeFont.java:433)
	at java.desktop/sun.font.GlyphList.mapChars(GlyphList.java:250)
	at java.desktop/sun.font.GlyphList.setFromString(GlyphList.java:225)
	at java.desktop/sun.java2d.pipe.GlyphListPipe.drawString(GlyphListPipe.java:70)
	at java.desktop/sun.java2d.SunGraphics2D.drawString(SunGraphics2D.java:2931)
	at ImageWriterCompressionTest.main(ImageWriterCompressionTest.java:82)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:586)
	at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
	at java.base/java.lang.Thread.run(Thread.java:1595)
```

@keithc-ca Do you have any feedback for this PR?

keithc-ca commented on runtime/j9vm/javanextvmi.cpp:
```diff
@@ -274,21 +274,56 @@ enterVThreadTransitionCritical(J9VMThread *currentThread, jobject thread)
 
-	while(!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(currentThread, threadObj, vm->virtualThreadInspectorCountOffset, 0, (U_64)-1)) {
+retry:
+	while (!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(currentThread, threadObj, vm->virtualThreadInspectorCountOffset, 0, (U_64)-1)) {
```
Directly casting a negative number to an unsigned type is a common anti-pattern in OpenJ9: a better expression than `(U_64)-1` would be `~(U_64)0`.

keithc-ca commented on runtime/jvmti/jvmtiForceEarlyReturn.cpp:
```diff
@@ -182,23 +188,23 @@ jvmtiForceEarlyReturn(jvmtiEnv* env,
-  							signatureType = 'L';
-  						}
-  						switch(signatureType) {
+				rc = (jvmtiError)findDecompileInfo(currentThread, threadToWalk, 0, &walkState);
```
This should include an intermediate cast through `(IDATA)` (don't change signedness and the width in a single cast).

fengxue-IS replied:
I thought `jvmtiError` is mapped to a `U_32` type (minimum 32 bit enum) https://github.com/eclipse-openj9/openj9/blob/master/debugtools/DDR_VM/src/com/ibm/j9ddr/StructureAliases29-edg.dat#L146, there shouldn't be any issues with direct casting from a `UDATA` to `jvmtiError`?

keithc-ca replied:
That file is only relevant for DDR (and then only in IBM builds). Here, we're dealing with the C++ compiler which only sees `typedef enum jvmtiError { ... } jvmtiError;` (see [`jtmti.h.m4`](https://github.com/eclipse-openj9/openj9/blob/master/runtime/include/jvmti.h.m4#L310)). Unless otherwise specified, the underlying type for an `enum` is `int` (see https://en.cppreference.com/w/c/language/enum). So, as written, this casts `UDATA` to (effectively) `int` which is both a width change and a signedness change: the cast through `IDATA` is needed.

keithc-ca commented on runtime/jvmti/jvmtiForceEarlyReturn.cpp:
```diff
@@ -182,23 +188,23 @@ jvmtiForceEarlyReturn(jvmtiEnv* env,
+				if (JVMTI_ERROR_NONE == rc) {
+					J9Method *method = (J9Method *)walkState.userData3;
+					J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(method);
+					if (romMethod->modifiers & J9AccNative) {
```
This should use `J9_ARE_ANY_BITS_SET()`.

keithc-ca commented on runtime/jvmti/jvmtiStackFrame.cpp:
```diff
@@ -490,7 +496,7 @@ jvmtiPopFrame(jvmtiEnv* env,
 				walkState.flags = J9_STACKWALK_INCLUDE_CALL_IN_FRAMES | J9_STACKWALK_INCLUDE_NATIVES | J9_STACKWALK_VISIBLE_ONLY | J9_STACKWALK_ITERATE_FRAMES | J9_STACKWALK_MAINTAIN_REGISTER_MAP;
 				vm->walkStackFrames(currentThread, &walkState);
-				rc = (UDATA)walkState.userData1;
+				rc = (jvmtiError)(UDATA)walkState.userData1;
```
UDATA -> IDATA

keithc-ca commented on runtime/jvmti/jvmtiStackFrame.cpp:
```diff
@@ -647,7 +653,7 @@ jvmtiNotifyFramePop(jvmtiEnv *env,
 #endif /* JAVA_SPEC_VERSION >= 19 */
 
-				rc = findDecompileInfo(currentThread, threadToWalk, (UDATA)depth, &walkState);
+				rc = (jvmtiError)findDecompileInfo(currentThread, threadToWalk, (UDATA)depth, &walkState);
```
Missing cast through `(IDATA)`.

keithc-ca commented on runtime/jvmti/jvmtiThread.cpp:
```diff
@@ -1193,16 +1214,15 @@ resumeThread(J9VMThread *currentThread, jthread thread)
-		releaseVMThread(currentThread, targetThread, thread);
 	}
 
+
```
Please don't add this duplicate blank line.

keithc-ca commented on runtime/jvmti/jvmtiThread.cpp:
```diff
@@ -1382,27 +1402,33 @@ jvmtiSuspendResumeCallBack(J9VMThread *vmThread, J9MM_IterateObjectDescriptor *o
 			j9object_t carrierThread = (j9object_t)J9VMJAVALANGVIRTUALTHREAD_CARRIERTHREAD(vmThread, vthread);
+			U_64 isSuspendedFlag = J9OBJECT_U64_LOAD(vmThread, vthread, vm->internalSuspendStateOffset);
+			J9VMThread *carrierVMThread = VM_VMHelpers::getCarrierVMThread(vmThread, vthread);
+			BOOLEAN inTransition = (NULL != carrierVMThread);
```
Here in C++ code, we should prefer `bool` over `BOOLEAN`.

keithc-ca commented on runtime/oti/VMHelpers.hpp:
```diff
@@ -2109,6 +2109,20 @@ class VM_VMHelpers
 		}
 	}
+
+	static VMINLINE BOOLEAN
```
Again, `bool` is a better choice than `BOOLEAN` in C++ code.

keithc-ca commented on runtime/oti/j9consts.h:
```diff
@@ -648,6 +648,12 @@ extern "C" {
+#if JAVA_SPEC_VERSION >= 19
+#define J9_VIRTUALTHREAD_INTERNAL_STATE_NONE      0x0
+#define J9_VIRTUALTHREAD_INTERNAL_STATE_SUSPENDED 0x1
+#define J9_VIRTUALTHREAD_INTERNAL_STATE_CARRIERID_MASK (~(uintptr_t)0xff)
```
Do we know that `J9VMThread` objects are allocated on 256 byte boundaries?
Since it seems we're only using the least significant bit for the state, the mask could be `(~(uintptr_t)0x1)`.

fengxue-IS replied:
All J9VMThread are aligned using `J9VMTHREAD_ALIGNMENT` which is `0x100`
Using `0xff` would indicate that the rest bits are reserved and safe for future use.

keithc-ca replied:
Thank you. A comment relating `J9_VIRTUALTHREAD_INTERNAL_STATE_CARRIERID_MASK` and `J9VMTHREAD_ALIGNMENT` would be welcome, or even
```c
#define J9_VIRTUALTHREAD_INTERNAL_STATE_CARRIERID_MASK (~(uintptr_t)(J9VMTHREAD_ALIGNMENT - 1))
```
but then `J9_VIRTUALTHREAD_INTERNAL_STATE_xxx` ought to move to `j9nonbuilder.h`.

keithc-ca commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -536,6 +536,22 @@ acquireVThreadInspector(J9VMThread *currentThread, jobject thread, BOOLEAN spin)
+			 * Decrement the inspector count to represent inspector during suspended transition.
+			 */
+			if (!objectAccessBarrier.inlineMixedObjectCompareAndSwapU64(
+															currentThread,
+															threadObj,
+															vm->virtualThreadInspectorCountOffset,
+															(U_64)vthreadInspectorCount,
+															((U_64)(vthreadInspectorCount - 1)))
```
The indentation is excessive here (two tabs more than line 543 is sufficient).
Also near lines 595 and 608.

fengxue-IS commented:
Updated code to address review comments

keithc-ca commented on runtime/oti/VMHelpers.hpp:
```diff
@@ -2109,6 +2109,40 @@ class VM_VMHelpers
 		}
 	}
+
+	/**
+	 * Check if thread is in suspended state.
+	 *
+	 * Current thread must have VM access.
+	 *
+	 * @param[in] currentThread the current J9VMThread
+	 * @param[in] thread the target Thread object
+	 * @return true if thread is suspend, false otherwise.
+	 *
+	 */
```
A article is missing in the summary, i.e. "the"
```
	 * Check if thread is in the suspended state.
```
In the `@return` clause, "suspend" should be "suspended". By convention, the clause does not end with a period, and line 2121 should be removed.
```
	 * @return true if thread is suspended, false otherwise
	 */
```

keithc-ca commented on runtime/oti/VMHelpers.hpp:
```diff
@@ -2109,6 +2109,40 @@ class VM_VMHelpers
+	 *
+	 * @param[in] currentThread the current J9VMThread
+	 * @param[in] thread the target vthread object
+	 * @return J9VMThread of the linked carrier VM thread, or NULL.
+	 *
```
Please remove the period and line 2138
```
	 * @return J9VMThread of the linked carrier VM thread, or NULL
```
(also "VM" seems redundant here)

keithc-ca reviewed (changes requested):
Please reformat the commit message(s) so no lines in the body are longer than 72 characters. Before merging, I expect this should be squashed - feel free to do that now.

keithc-ca commented on runtime/oti/j9consts.h:
```diff
@@ -648,6 +648,15 @@ extern "C" {
 
+#if JAVA_SPEC_VERSION >= 19
+/* J9VMThread address is aligned using J9VMTHREAD_ALIGNMENT (0x100)
+ * So last 8 bits can be used as bit flags.
```
So -> so

keithc-ca commented on runtime/jvmti/jvmtiStackFrame.cpp:
```diff
@@ -490,7 +496,7 @@ jvmtiPopFrame(jvmtiEnv* env,
 				walkState.flags = J9_STACKWALK_INCLUDE_CALL_IN_FRAMES | J9_STACKWALK_INCLUDE_NATIVES | J9_STACKWALK_VISIBLE_ONLY | J9_STACKWALK_ITERATE_FRAMES | J9_STACKWALK_MAINTAIN_REGISTER_MAP;
 				vm->walkStackFrames(currentThread, &walkState);
-				rc = (UDATA)walkState.userData1;
+				rc = (jvmtiError)(IDATA)(UDATA)walkState.userData1;
```
We don't need the cast through `UDATA` (`IDATA` is the same width as `void *`).

babsingh commented:
jenkins compile plinux jdk8

babsingh commented:
jenkins test extended.openjdk aix jdk21

babsingh commented:
jenkins test extended.openjdk alinux jdk22

babsingh reviewed (approved):
LGTM. Known `ImageWriterCompressionTest` and `DatagramChannel` failures are seen in the `extended.openjdk` PR builds, which are unrelated to this PR.

pshipton commented:
I'm open to it, if the testing today/tonight goes well. It seems there is little risk to any versions other than 21, 22, and mainly virtual threads are affected.

pshipton commented:
Pls go ahead with the backport.
